### PR TITLE
Have metrics respect mask and sample weight

### PR DESCRIPTION
**Pull request by carlthome**
If a model uses masking, some metrics will not be calculated correctly. For example, a mask_value > 1 with precision.

```py
import numpy as np
from keras.models import *
from keras.layers import *

nb_samples = 32
nb_timesteps = 10
nb_features = 15
nb_hidden = 10
nb_labels = 5
mask_value = 2

# Generate some random data. Mask out some sequences with mask_value.
x = np.random.rand(nb_samples, nb_timesteps, nb_features)
y = np.random.rand(nb_samples, nb_timesteps, nb_labels)
x[1:, 5:] = y[1:, 5:] = mask_value

i = Input((nb_timesteps, nb_features))
o = Masking(mask_value)(i)
o = LSTM(nb_hidden, return_sequences=True)(o)
o = TimeDistributed(Dense(nb_labels, activation='sigmoid'))(o)
model = Model(i, o)
model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['precision'])
model.fit(x, y, batch_size=nb_samples, shuffle=False)
```

>  Epoch 1/10
32/32 [==============================] - 0s - loss: 0.7058 - precision: 1.0807
Epoch 2/10
32/32 [==============================] - 0s - loss: 0.7033 - precision: 1.1125
Epoch 3/10
32/32 [==============================] - 0s - loss: 0.7018 - precision: 1.1253
Epoch 4/10
32/32 [==============================] - 0s - loss: 0.7007 - precision: 1.1681
Epoch 5/10
32/32 [==============================] - 0s - loss: 0.6998 - precision: 1.2192
Epoch 6/10
32/32 [==============================] - 0s - loss: 0.6991 - precision: 1.1958
Epoch 7/10
32/32 [==============================] - 0s - loss: 0.6984 - precision: 1.2187
Epoch 8/10
32/32 [==============================] - 0s - loss: 0.6978 - precision: 1.2414
Epoch 9/10
32/32 [==============================] - 0s - loss: 0.6973 - precision: 1.2720
Epoch 10/10
32/32 [==============================] - 0s - loss: 0.6969 - precision: 1.2383

# TODO
 - [ ] Add unit test
 - [ ] Confirm that metrics should respect masking and sample weights. @fchollet, am I missing something?
 - [ ] Decorate accuracy metrics with weighted_objective too.

### CODE REVIEW

**Comment by carlthome** [comment]
I have some trouble with `utils.generic_utils.get_from_module`, hence failed checks when metrics are dicts (multi-output models). If anyone feels like doing some refactoring to prepare metrics like losses in training.py that would be sweet. Otherwise I'll look into this again during the weekend.

**Comment by fchollet** [comment]
All that would be needed would be to reproduce the same behavior as the losses processing. Agreed that's a good thing to do.

**Comment by carlthome** [comment]
I've been looking at this for an hour now, but honestly, should it be necessary to understand the convoluted mess that is `get_from_module` to achieve this? I fear using Keras is taking up more development time than a good code base would, and I'm having doubts Keras will ever mature into a stable package without major refactoring from more experienced developers. There's a lot of code smell and I've thought most of the complexity comes from too many additions and not enough deletions but looking through `git blame` I've noticed most of the bad design choices stem from the initial release.

If I were you I'd consider documenting the code base properly (e.g. every function should have a docstring), and removing a lot of reinvented boilerplate. For example: all of `generic_utils.py` should be replaced entirely with libraries; you should use `abc` instead of rolling your own abstract base classes; all the type checking in every function bloats the code base; there's code duplication everywhere (`objectives.py` vs. `metrics.py`) and misplaced logic (e.g. accuracy handling in `training.py`); I really don't see why `globals()` is used at all to maintain state in Keras, and it's all very non-pythonic; please settle on a docstring convention (see Lasagne);

There will be more bugs as development progresses so I'm abandoning the ship for [TFLearn](https://github.com/tflearn/tflearn) and [Lasagne](https://github.com/Lasagne/Lasagne) as of now. I really love Keras's API and appreciate your initial efforts, @fchollet, but there's too much instability and poor software engineering.
